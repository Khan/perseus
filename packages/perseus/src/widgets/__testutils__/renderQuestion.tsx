import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
