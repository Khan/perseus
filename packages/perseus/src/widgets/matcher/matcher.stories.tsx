import {ServerItemRendererWithDebugUI} from "../../../../../testing/server-item-renderer-with-debug-ui";
import {generateTestPerseusItem} from "../../util/test-utils";

import {question1} from "./matcher.testdata";

import type {Meta, StoryObj} from "@storybook/react";

const meta: Meta = {
    title: "Perseus/Widgets/Matcher",
    component: ServerItemRendererWithDebugUI,
};
export default meta;

type Story = StoryObj<typeof ServerItemRendererWithDebugUI>;

export const Question1: Story = {
    args: {
        item: generateTestPerseusItem({question: question1}),
    },
};
