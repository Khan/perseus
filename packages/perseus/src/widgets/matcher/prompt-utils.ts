import type matcher from "./matcher";
import type {WidgetType} from "../../prompt-types";
import type {PerseusMatcherUserInput} from "../../validation.types";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type WidgetProps = PropsFor<typeof matcher.widget>;

export type MatcherPromptJSON = {
    type: WidgetType;
    options: {
        labels: WidgetProps["labels"];
        left: WidgetProps["left"];
        right: WidgetProps["right"];
        orderMatters: WidgetProps["orderMatters"];
    };
    userInput: {
        left: PerseusMatcherUserInput["left"];
        right: PerseusMatcherUserInput["right"];
    };
};

export const getPromptJSON = (
    renderProps: WidgetProps,
    userInput: PerseusMatcherUserInput,
): MatcherPromptJSON => {
    return {
        type: "matcher",
        options: {
            labels: renderProps.labels,
            left: renderProps.left,
            right: renderProps.right,
            orderMatters: renderProps.orderMatters,
        },
        userInput: {
            left: userInput.left,
            right: userInput.right,
        },
    };
};
