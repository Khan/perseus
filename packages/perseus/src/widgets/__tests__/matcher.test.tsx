import * as React from "react";
import "@testing-library/jest-dom";

import {testDependencies} from "../../../../../testing/test-dependencies";
import {wait} from "../../../../../testing/wait";
import * as Dependencies from "../../dependencies";
import {question1} from "../__testdata__/matcher.testdata";

import {renderQuestion} from "./renderQuestion";

import type {APIOptions} from "../../types";
import type {Matcher} from "../matcher";

describe("matcher widget", () => {
    beforeEach(() => {
        /*
        Sortable misbehaves and sets state after the component has been
        unmounted. This is existing behavior and its safer to leave the existing
        implementation and swallow the warning in tests.
        */
        jest.spyOn(console, "warn").mockImplementation(() => {});
        jest.spyOn(console, "error").mockImplementation(() => {});

        jest.useRealTimers();

        jest.spyOn(Dependencies, "getDependencies").mockReturnValue({
            ...testDependencies,
            TeX: ({
                children,
                onRender: onLoad,
            }: {
                children: React.ReactNode;
                onRender?: () => unknown;
            }) => {
                React.useLayoutEffect(() => {
                    onLoad?.();
                }, [onLoad]);
                return <span className="tex-mock">{children}</span>;
            },
        });
    });

    it("should snapshot", async () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };

        // Act
        const {container} = renderQuestion(question1, apiOptions);
        await wait();

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should snapshot on mobile", async () => {
        // Arrange
        jest.useRealTimers();

        const apiOptions: APIOptions = {
            isMobile: true,
        };

        // Act
        const {container} = renderQuestion(question1, apiOptions);
        await wait();

        // Assert
        expect(container).toMatchSnapshot("first mobile render");
    });

    it("can reorder options", async () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {container, renderer} = renderQuestion(question1, apiOptions);
        await wait();

        // Act
        const matcher: Matcher = renderer.findWidgets("matcher 1")[0];

        matcher.moveRightOptionToIndex(
            "Rapid escalation of greenhouse gas emissions",
            0,
        );

        matcher.moveLeftOptionToIndex(
            "Average global temperatures will rise ",
            0,
        );

        // Assert
        expect(container).toMatchSnapshot("moved items");
    });

    it("can be answered correctly", async () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);
        await wait();

        const matcher: Matcher = renderer.findWidgets("matcher 1")[0];

        // Put the right options in the correct order by repeatedly moving
        // answers to the end of the list
        [
            "Medium-sized stars typically exist for roughly 10 billion years",
            "The current trajectory of the Earth\u2019s tectonic plate movement",
            "The life cycle of medium-sized stars includes a red giant stage and ends in a whimper as a white dwarf",
            "Rapid escalation of greenhouse gas emissions",
            "The current trajectory of the Milky Way galaxy and those in its immediate proximity",
        ].forEach((option, index) => {
            matcher.moveRightOptionToIndex(option, 4);
        });

        // Act
        renderer.guessAndScore();

        // assert
        expect(renderer).toHaveBeenAnsweredCorrectly();
    });

    it("can be answered incorrectly", async () => {
        // Arrange
        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);
        await wait();

        const matcher: Matcher = renderer.findWidgets("matcher 1")[0];

        // Put the left options in reverse order
        [
            "Our Sun will run out of fuel and die in around 5 billion years ",
            "Plate tectonics will rearrange the continents: the Pacific will narrow, bringing Australia closer to the Americas, and the Atlantic will expand to form the largest of the oceans ",
            "Our Sun will run out of hydrogen, swell into a red giant, gobble up the inner rocky planets, and then collapse and die ",
            "Average global temperatures will rise ",
            "In 3 to 4 billion years, our galaxy will begin a slow collision with its closest large neighbor, Andromeda ",
        ].forEach((option, index) => {
            matcher.moveLeftOptionToIndex(option, 0);
        });

        // Act
        renderer.guessAndScore();

        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });

    it("is scored incorrect if the math renderer hasn't loaded yet", () => {
        // Arrange: stub the TeX renderer to never call its onRender prop,
        // which is how the matcher knows the math renderer is ready.
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue({
            ...testDependencies,
            TeX: () => {
                return null;
            },
        });

        const apiOptions: APIOptions = {
            isMobile: false,
        };
        const {renderer} = renderQuestion(question1, apiOptions);

        // Act
        renderer.guessAndScore();

        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });
});
