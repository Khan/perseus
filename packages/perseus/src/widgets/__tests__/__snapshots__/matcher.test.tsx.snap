// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matcher widget can reorder options: moved items 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <strong>
          Match each claim with its supporting evidence.
        </strong>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <table
            class="widget_lf609f perseus-widget-matcher"
          >
            <tbody>
              <tr
                class="row_tykfnp"
              >
                <th
                  class="column_4at8du-o_O-columnLabel_1wdbfym"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        <strong>
                          Claims
                        </strong>
                      </div>
                    </div>
                  </div>
                </th>
                <th
                  class="column_4at8du-o_O-columnRight_1wcmc4d-o_O-columnLabel_1wdbfym"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        <strong>
                          Evidence
                        </strong>
                      </div>
                    </div>
                  </div>
                </th>
              </tr>
              <tr
                class="row_tykfnp"
              >
                <td
                  class="column_4at8du"
                >
                  <ul
                    class="sortable_13d4756 perseus-sortable"
                  >
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Average global temperatures will rise 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Our Sun will run out of fuel and die in around 5 billion years 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Plate tectonics will rearrange the continents: the Pacific will narrow, bringing Australia closer to the Americas, and the Atlantic will expand to form the largest of the oceans 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Our Sun will run out of hydrogen, swell into a red giant, gobble up the inner rocky planets, and then collapse and die 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            In 3 to 4 billion years, our galaxy will begin a slow collision with its closest large neighbor, Andromeda 
                          </div>
                        </div>
                      </div>
                    </li>
                  </ul>
                </td>
                <td
                  class="column_4at8du-o_O-columnRight_1wcmc4d"
                >
                  <ul
                    class="sortable_13d4756 perseus-sortable"
                  >
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Rapid escalation of greenhouse gas emissions
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The current trajectory of the Milky Way galaxy and those in its immediate proximity
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Medium-sized stars typically exist for roughly 10 billion years
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The life cycle of medium-sized stars includes a red giant stage and ends in a whimper as a white dwarf
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The current trajectory of the Earth’s tectonic plate movement
                          </div>
                        </div>
                      </div>
                    </li>
                  </ul>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`matcher widget should snapshot on mobile: first mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <strong>
          Match each claim with its supporting evidence.
        </strong>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <table
            class="widget_lf609f perseus-widget-matcher"
          >
            <tbody>
              <tr
                class="row_tykfnp"
              >
                <th
                  class="column_4at8du-o_O-columnLabel_1wdbfym"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        <strong>
                          Claims
                        </strong>
                      </div>
                    </div>
                  </div>
                </th>
                <th
                  class="column_4at8du-o_O-columnRight_1wcmc4d-o_O-columnLabel_1wdbfym"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        <strong>
                          Evidence
                        </strong>
                      </div>
                    </div>
                  </div>
                </th>
              </tr>
              <tr
                class="row_tykfnp"
              >
                <td
                  class="column_4at8du"
                >
                  <ul
                    class="sortable_13d4756 perseus-sortable"
                  >
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 8px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Our Sun will run out of fuel and die in around 5 billion years 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 8px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Plate tectonics will rearrange the continents: the Pacific will narrow, bringing Australia closer to the Americas, and the Atlantic will expand to form the largest of the oceans 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 8px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Our Sun will run out of hydrogen, swell into a red giant, gobble up the inner rocky planets, and then collapse and die 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 8px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Average global temperatures will rise 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            In 3 to 4 billion years, our galaxy will begin a slow collision with its closest large neighbor, Andromeda 
                          </div>
                        </div>
                      </div>
                    </li>
                  </ul>
                </td>
                <td
                  class="column_4at8du-o_O-columnRight_1wcmc4d"
                >
                  <ul
                    class="sortable_13d4756 perseus-sortable"
                  >
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 8px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The current trajectory of the Milky Way galaxy and those in its immediate proximity
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 8px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Medium-sized stars typically exist for roughly 10 billion years
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 8px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The life cycle of medium-sized stars includes a red giant stage and ends in a whimper as a white dwarf
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 8px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The current trajectory of the Earth’s tectonic plate movement
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Rapid escalation of greenhouse gas emissions
                          </div>
                        </div>
                      </div>
                    </li>
                  </ul>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`matcher widget should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <strong>
          Match each claim with its supporting evidence.
        </strong>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <table
            class="widget_lf609f perseus-widget-matcher"
          >
            <tbody>
              <tr
                class="row_tykfnp"
              >
                <th
                  class="column_4at8du-o_O-columnLabel_1wdbfym"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        <strong>
                          Claims
                        </strong>
                      </div>
                    </div>
                  </div>
                </th>
                <th
                  class="column_4at8du-o_O-columnRight_1wcmc4d-o_O-columnLabel_1wdbfym"
                >
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        <strong>
                          Evidence
                        </strong>
                      </div>
                    </div>
                  </div>
                </th>
              </tr>
              <tr
                class="row_tykfnp"
              >
                <td
                  class="column_4at8du"
                >
                  <ul
                    class="sortable_13d4756 perseus-sortable"
                  >
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Our Sun will run out of fuel and die in around 5 billion years 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Plate tectonics will rearrange the continents: the Pacific will narrow, bringing Australia closer to the Americas, and the Atlantic will expand to form the largest of the oceans 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Our Sun will run out of hydrogen, swell into a red giant, gobble up the inner rocky planets, and then collapse and die 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Average global temperatures will rise 
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1-o_O-disabled_u0qq3k perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            In 3 to 4 billion years, our galaxy will begin a slow collision with its closest large neighbor, Andromeda 
                          </div>
                        </div>
                      </div>
                    </li>
                  </ul>
                </td>
                <td
                  class="column_4at8du-o_O-columnRight_1wcmc4d"
                >
                  <ul
                    class="sortable_13d4756 perseus-sortable"
                  >
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The current trajectory of the Milky Way galaxy and those in its immediate proximity
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Medium-sized stars typically exist for roughly 10 billion years
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The life cycle of medium-sized stars includes a red giant stage and ends in a whimper as a white dwarf
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px 0px 5px 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            The current trajectory of the Earth’s tectonic plate movement
                          </div>
                        </div>
                      </div>
                    </li>
                    <li
                      class="card_1s6cuhy-o_O-draggable_12to336-o_O-verticalCard_1gw9js1 perseus-interactive perseus-sortable-draggable"
                      style="position: static; margin: 0px;"
                    >
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            Rapid escalation of greenhouse gas emissions
                          </div>
                        </div>
                      </div>
                    </li>
                  </ul>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
